# Assignment 1: Tokenization & POS Tagging (Chinese)
**Date:** 2026-03-12

## Learning Objectives
- Perform sentence segmentation and tokenization on Chinese text
- Apply Part-of-Speech (POS) tagging
- Analyze token frequency and POS distributions

> **Note (說明)：** Use **Chinese text** as your main material. You may use built-in samples below.

## Allowed Libraries
- `re`, `collections`
- `jieba`, `jieba.posseg` (recommended for quick start)
- `spaCy` with a Chinese model (optional)
- `stanza` (optional)

## Submission Format: PDF  (繳交格式)
- We need to assess your assignment by your code and the results. You can export a jupyter notebook with results as a PDF file, or you can summarize your code and results in other software but eventually export as a PDF file.
- Remember: The file you submit should **contain your code and results**. (檔案必須為PDF檔，內容須可見你的程式碼以及執行結果)

## Deadline: 繳交期限
- 2026/03/27  08:30
- You have two weeks to finish this work

---

## 0. Setup

In [1]:
# 環境設定（使用 uv 虛擬環境）
# !uv pip install jieba spacy
# !uv pip install zh-core-web-sm@https://github.com/explosion/spacy-models/releases/download/zh_core_web_sm-3.8.0/zh_core_web_sm-3.8.0-py3-none-any.whl

import warnings
warnings.filterwarnings("ignore", category=SyntaxWarning)

## 1. Sample Data (Replace with your own if desired)

In [2]:
SENTENCES_ZH = [
    "今天是個好天氣，我們一起去台北市立美術館吧！",
    "自然語言處理在教育、醫療與金融領域都有廣泛應用。",
    "請問下週三下午兩點在BS440教室還有課嗎？",
    "你覺得大型語言模型在中文任務上的表現如何？",
    "他知道這件事不要緊",
    "這份報告我寫不好",
    "他有一個女兒，在醫院工作",
    "我在樓上看到了他",
    "兒的生活好痛苦也沒有糧食多病少掙了很多錢"
]

## 2. Sentence Segmentation (句子切分)

In [3]:
import re

def split_sentences(text: str):
    """Split Chinese text into sentences using punctuation heuristics."""
    parts = re.split(r"(?<=[。！？!?])", text)
    return [p.strip() for p in parts if p.strip()]

# TODO: Test your function on a paragraph of your own
for s in SENTENCES_ZH:
    print(split_sentences(s))

['今天是個好天氣，我們一起去台北市立美術館吧！']
['自然語言處理在教育、醫療與金融領域都有廣泛應用。']
['請問下週三下午兩點在BS440教室還有課嗎？']
['你覺得大型語言模型在中文任務上的表現如何？']
['他知道這件事不要緊']
['這份報告我寫不好']
['他有一個女兒，在醫院工作']
['我在樓上看到了他']
['兒的生活好痛苦也沒有糧食多病少掙了很多錢']


## 3. Tokenization (斷詞)

In [4]:
import jieba

def tokenize_zh(text: str):
    return [tok for tok in jieba.lcut(text) if tok.strip()]

for s in SENTENCES_ZH:
    print(tokenize_zh(s))

Building prefix dict from the default dictionary ...


Loading model from cache /var/folders/56/qpw4f8jd6m5gdyr1w3x_gn300000gn/T/jieba.cache


Loading model cost 0.389 seconds.


Prefix dict has been built successfully.


['今天', '是', '個', '好', '天氣', '，', '我們', '一起', '去', '台北市立', '美術館', '吧', '！']
['自然', '語言', '處理', '在', '教育', '、', '醫療', '與', '金融', '領域', '都', '有', '廣泛', '應用', '。']
['請', '問下', '週三', '下午', '兩點', '在', 'BS440', '教室', '還有', '課', '嗎', '？']
['你', '覺得', '大型', '語言', '模型', '在', '中文', '任務上', '的', '表現', '如何', '？']
['他', '知道', '這件', '事', '不要', '緊']
['這份', '報告', '我', '寫', '不好']
['他', '有', '一個', '女兒', '，', '在', '醫院', '工作']
['我', '在', '樓', '上', '看到', '了', '他']
['兒', '的', '生活', '好', '痛苦', '也', '沒', '有', '糧食', '多', '病少', '掙', '了', '很多', '錢']


## 4. POS Tagging (詞性標註)

In [5]:
from jieba import posseg as pseg

def pos_tag_zh(text: str):
    return [(w.word, w.flag) for w in pseg.cut(text)]

for s in SENTENCES_ZH:
    print(pos_tag_zh(s))

[('今天', 't'), ('是', 'v'), ('個', 'q'), ('好', 'a'), ('天氣', 'n'), ('，', 'x'), ('我們', 'r'), ('一起', 'm'), ('去', 'v'), ('台北市立', 'nt'), ('美術館', 'nt'), ('吧', 'y'), ('！', 'x')]
[('自然', 'd'), ('語言', 'n'), ('處理', 'v'), ('在', 'p'), ('教育', 'vn'), ('、', 'x'), ('醫療', 'n'), ('與', 'p'), ('金融', 'n'), ('領域', 'n'), ('都', 'd'), ('有', 'v'), ('廣泛', 'a'), ('應用', 'v'), ('。', 'x')]


[('請', 'v'), ('問', 'v'), ('下週', 't'), ('三', 'm'), ('下午', 't'), ('兩點', 'm'), ('在', 'p'), ('BS440', 'eng'), ('教室', 'n'), ('還', 'd'), ('有', 'v'), ('課', 'n'), ('嗎', 'y'), ('？', 'x')]
[('你', 'r'), ('覺得', 'v'), ('大型', 'b'), ('語言', 'n'), ('模型', 'n'), ('在', 'p'), ('中文', 'nz'), ('任務', 'n'), ('上', 'f'), ('的', 'uj'), ('表現', 'v'), ('如何', 'r'), ('？', 'x')]
[('他', 'r'), ('知道', 'v'), ('這件', 'mq'), ('事', 'n'), ('不要', 'df'), ('緊', 'x')]
[('這份', 'mq'), ('報告', 'n'), ('我', 'r'), ('寫', 'v'), ('不好', 'd')]
[('他', 'r'), ('有', 'v'), ('一個', 'm'), ('女兒', 'n'), ('，', 'x'), ('在', 'p'), ('醫院', 'n'), ('工作', 'vn')]
[('我', 'r'), ('在', 'p'), ('樓上', 's'), ('看到', 'v'), ('了', 'ul'), ('他', 'r')]
[('兒', 'n'), ('的', 'uj'), ('生活', 'vn'), ('好', 'a'), ('痛苦', 'an'), ('也', 'd'), ('沒有', 'v'), ('糧食', 'n'), ('多', 'm'), ('病', 'n'), ('少', 'a'), ('掙', 'v'), ('了', 'ul'), ('很多', 'm'), ('錢', 'x')]


## 5. Analysis: Frequency & POS Distribution

In [6]:
from collections import Counter

all_tokens = []
all_pos = []
for s in SENTENCES_ZH:
    toks = tokenize_zh(s)
    all_tokens.extend(toks)
    all_pos.extend([p for _, p in pos_tag_zh(s)])

print('Top tokens:', Counter(all_tokens).most_common(10))
print('POS distribution:', Counter(all_pos).most_common())

Top tokens: [('在', 5), ('有', 3), ('他', 3), ('好', 2), ('，', 2), ('語言', 2), ('？', 2), ('的', 2), ('我', 2), ('了', 2)]
POS distribution: [('n', 17), ('v', 16), ('x', 9), ('r', 8), ('m', 6), ('p', 6), ('d', 5), ('a', 4), ('t', 3), ('vn', 3), ('nt', 2), ('y', 2), ('uj', 2), ('mq', 2), ('ul', 2), ('q', 1), ('eng', 1), ('b', 1), ('nz', 1), ('f', 1), ('df', 1), ('s', 1), ('an', 1)]


## 6. 比較 jieba 與 spaCy 的斷詞結果

我先用上面的 sample code 跑完 jieba 的斷詞、POS tagging 和詞頻統計，確認都能正常執行後，接著用 spaCy 的 `zh_core_web_sm` 中文模型對同一批句子做比較，看看兩套工具在實際結果上有什麼差異。

### 6.1 用 spaCy 跑同一批句子

In [7]:
import spacy

nlp = spacy.load("zh_core_web_sm")

def tokenize_spacy(text: str):
    doc = nlp(text)
    return [tok.text for tok in doc if tok.text.strip()]

def pos_tag_spacy(text: str):
    doc = nlp(text)
    return [(tok.text, tok.pos_) for tok in doc if tok.text.strip()]

print("=== spaCy Tokenization ===")
for s in SENTENCES_ZH:
    print(tokenize_spacy(s))

print("\n=== spaCy POS Tagging ===")
for s in SENTENCES_ZH:
    print(pos_tag_spacy(s))

=== spaCy Tokenization ===
['今天', '是', '個好', '天氣', '，', '我', '們', '一起', '去', '台北市', '立美', '術館', '吧', '！']
['自然', '語言', '處理', '在', '教育', '、', '醫療', '與', '金融', '領域', '都', '有', '廣泛', '應用', '。']
['請問', '下週', '三', '下午', '兩點', '在', 'BS', '440', '教室', '還', '有', '課嗎', '？']
['你', '覺', '得', '大型', '語言', '模型', '在', '中文', '任務', '上', '的', '表現', '如何', '？']
['他', '知道', '這件', '事', '不要', '緊']
['這份', '報告', '我', '寫', '不', '好']
['他', '有', '一', '個', '女兒', '，', '在', '醫院', '工作']
['我', '在', '樓上', '看到', '了', '他']
['兒', '的', '生活', '好', '痛苦', '也', '沒有', '糧食', '多', '病少', '掙', '了', '很多', '錢']

=== spaCy POS Tagging ===
[('今天', 'NOUN'), ('是', 'VERB'), ('個好', 'NOUN'), ('天氣', 'VERB'), ('，', 'PUNCT'), ('我', 'PRON'), ('們', 'VERB'), ('一起', 'ADV'), ('去', 'VERB'), ('台北市', 'PROPN'), ('立美', 'VERB'), ('術館', 'VERB'), ('吧', 'PART'), ('！', 'PUNCT')]
[('自然', 'ADV'), ('語言', 'VERB'), ('處理', 'VERB'), ('在', 'ADP'), ('教育', 'NOUN'), ('、', 'PUNCT'), ('醫療', 'NOUN'), ('與', 'VERB'), ('金融', 'NOUN'), ('領域', 'NOUN'), ('都', 'ADV'), ('有', 'VERB

### 6.2 逐句比較斷詞結果

接下來把兩套工具的斷詞結果放在一起對照，比較容易看出差異在哪裡。

In [8]:
print("=" * 80)
print("jieba vs spaCy — 斷詞比較")
print("=" * 80)

for i, s in enumerate(SENTENCES_ZH):
    jieba_toks = tokenize_zh(s)
    spacy_toks = tokenize_spacy(s)
    print(f"\n--- 句子 {i+1}: {s}")
    print(f"  jieba ({len(jieba_toks)} tokens): {' / '.join(jieba_toks)}")
    print(f"  spaCy ({len(spacy_toks)} tokens): {' / '.join(spacy_toks)}")
    
    # 找出差異
    if jieba_toks != spacy_toks:
        jieba_set = set(jieba_toks)
        spacy_set = set(spacy_toks)
        only_jieba = jieba_set - spacy_set
        only_spacy = spacy_set - jieba_set
        if only_jieba:
            print(f"  → jieba 獨有: {only_jieba}")
        if only_spacy:
            print(f"  → spaCy 獨有: {only_spacy}")

jieba vs spaCy — 斷詞比較

--- 句子 1: 今天是個好天氣，我們一起去台北市立美術館吧！
  jieba (13 tokens): 今天 / 是 / 個 / 好 / 天氣 / ， / 我們 / 一起 / 去 / 台北市立 / 美術館 / 吧 / ！
  spaCy (14 tokens): 今天 / 是 / 個好 / 天氣 / ， / 我 / 們 / 一起 / 去 / 台北市 / 立美 / 術館 / 吧 / ！
  → jieba 獨有: {'美術館', '我們', '個', '台北市立', '好'}
  → spaCy 獨有: {'個好', '我', '們', '台北市', '術館', '立美'}

--- 句子 2: 自然語言處理在教育、醫療與金融領域都有廣泛應用。
  jieba (15 tokens): 自然 / 語言 / 處理 / 在 / 教育 / 、 / 醫療 / 與 / 金融 / 領域 / 都 / 有 / 廣泛 / 應用 / 。
  spaCy (15 tokens): 自然 / 語言 / 處理 / 在 / 教育 / 、 / 醫療 / 與 / 金融 / 領域 / 都 / 有 / 廣泛 / 應用 / 。

--- 句子 3: 請問下週三下午兩點在BS440教室還有課嗎？
  jieba (12 tokens): 請 / 問下 / 週三 / 下午 / 兩點 / 在 / BS440 / 教室 / 還有 / 課 / 嗎 / ？
  spaCy (13 tokens): 請問 / 下週 / 三 / 下午 / 兩點 / 在 / BS / 440 / 教室 / 還 / 有 / 課嗎 / ？
  → jieba 獨有: {'還有', '週三', '問下', '請', 'BS440', '課', '嗎'}
  → spaCy 獨有: {'有', '請問', '還', '440', '下週', 'BS', '課嗎', '三'}

--- 句子 4: 你覺得大型語言模型在中文任務上的表現如何？
  jieba (12 tokens): 你 / 覺得 / 大型 / 語言 / 模型 / 在 / 中文 / 任務上 / 的 / 表現 / 如何 / ？
  spaCy (14 tokens): 你 / 覺 / 得 / 大型 / 語言 / 模型 / 在 / 中文 

### 6.3 詞性標註比較

兩套工具用的 POS 標籤系統完全不同，jieba 用北大標註集（像 `n`, `v`, `r`），spaCy 用 Universal POS Tags（像 `NOUN`, `VERB`, `PRON`），所以不能直接比標籤名稱，但可以看各自標得合不合理。

In [9]:
print("=" * 60)
print("jieba vs spaCy — 詞性標註比較")
print("=" * 60)

for i, s in enumerate(SENTENCES_ZH):
    jieba_pos = pos_tag_zh(s)
    spacy_pos = pos_tag_spacy(s)
    print(f"\n--- 句子 {i+1}: {s}")
    print(f"  jieba : {jieba_pos}")
    print(f"  spaCy : {spacy_pos}")

jieba vs spaCy — 詞性標註比較

--- 句子 1: 今天是個好天氣，我們一起去台北市立美術館吧！
  jieba : [('今天', 't'), ('是', 'v'), ('個', 'q'), ('好', 'a'), ('天氣', 'n'), ('，', 'x'), ('我們', 'r'), ('一起', 'm'), ('去', 'v'), ('台北市立', 'nt'), ('美術館', 'nt'), ('吧', 'y'), ('！', 'x')]
  spaCy : [('今天', 'NOUN'), ('是', 'VERB'), ('個好', 'NOUN'), ('天氣', 'VERB'), ('，', 'PUNCT'), ('我', 'PRON'), ('們', 'VERB'), ('一起', 'ADV'), ('去', 'VERB'), ('台北市', 'PROPN'), ('立美', 'VERB'), ('術館', 'VERB'), ('吧', 'PART'), ('！', 'PUNCT')]



--- 句子 2: 自然語言處理在教育、醫療與金融領域都有廣泛應用。
  jieba : [('自然', 'd'), ('語言', 'n'), ('處理', 'v'), ('在', 'p'), ('教育', 'vn'), ('、', 'x'), ('醫療', 'n'), ('與', 'p'), ('金融', 'n'), ('領域', 'n'), ('都', 'd'), ('有', 'v'), ('廣泛', 'a'), ('應用', 'v'), ('。', 'x')]
  spaCy : [('自然', 'ADV'), ('語言', 'VERB'), ('處理', 'VERB'), ('在', 'ADP'), ('教育', 'NOUN'), ('、', 'PUNCT'), ('醫療', 'NOUN'), ('與', 'VERB'), ('金融', 'NOUN'), ('領域', 'NOUN'), ('都', 'ADV'), ('有', 'VERB'), ('廣泛', 'NOUN'), ('應用', 'VERB'), ('。', 'PUNCT')]



--- 句子 3: 請問下週三下午兩點在BS440教室還有課嗎？
  jieba : [('請', 'v'), ('問', 'v'), ('下週', 't'), ('三', 'm'), ('下午', 't'), ('兩點', 'm'), ('在', 'p'), ('BS440', 'eng'), ('教室', 'n'), ('還', 'd'), ('有', 'v'), ('課', 'n'), ('嗎', 'y'), ('？', 'x')]
  spaCy : [('請問', 'VERB'), ('下週', 'NOUN'), ('三', 'NUM'), ('下午', 'NOUN'), ('兩點', 'VERB'), ('在', 'ADP'), ('BS', 'PROPN'), ('440', 'NUM'), ('教室', 'NOUN'), ('還', 'VERB'), ('有', 'VERB'), ('課嗎', 'NOUN'), ('？', 'PUNCT')]

--- 句子 4: 你覺得大型語言模型在中文任務上的表現如何？
  jieba : [('你', 'r'), ('覺得', 'v'), ('大型', 'b'), ('語言', 'n'), ('模型', 'n'), ('在', 'p'), ('中文', 'nz'), ('任務', 'n'), ('上', 'f'), ('的', 'uj'), ('表現', 'v'), ('如何', 'r'), ('？', 'x')]
  spaCy : [('你', 'PRON'), ('覺', 'VERB'), ('得', 'PART'), ('大型', 'ADJ'), ('語言', 'NOUN'), ('模型', 'NOUN'), ('在', 'ADP'), ('中文', 'NOUN'), ('任務', 'NOUN'), ('上', 'PART'), ('的', 'PART'), ('表現', 'NOUN'), ('如何', 'VERB'), ('？', 'PUNCT')]

--- 句子 5: 他知道這件事不要緊
  jieba : [('他', 'r'), ('知道', 'v'), ('這件', 'mq'), ('事', 'n'), ('不要', 'df'), ('緊', 'x')]
  spaCy : [('他', 

### 6.4 觀察與分析

跑完之後我覺得蠻意外的，一開始以為 spaCy 用了深度學習模型應該會比 jieba 穩，但實際上在這組中文句子裡，spaCy 的斷詞反而切得比較碎、比較不自然。以下挑幾個差異比較明顯的例子來討論，有些句子兩套工具結果差不多就不展開了。

---

**例子 1：「今天是個好天氣，我們一起去台北市立美術館吧！」**

- jieba：`台北市立 / 美術館` → 雖然沒有完整保留「台北市立美術館」，但至少還看得出是一個地名
- spaCy：`台北市 / 立美 / 術館` → 「立美」「術館」完全不是詞，切得很離譜
- 另外 spaCy 還把「個好」黏在一起，把「我們」拆成「我 / 們」，這些都不太合理

在專有名詞的處理上，jieba 明顯比 spaCy 好很多。

---

**例子 2：「請問下週三下午兩點在BS440教室還有課嗎？」**

- jieba：把「BS440」當成一個詞，但出現了「問下」這種不自然的組合
- spaCy：把「BS440」拆成「BS / 440」，然後「課嗎」黏在一起
- 兩套工具都沒辦法完美處理這種中英數混合的句子，各有各的問題

---

**例子 3：「這份報告我寫不好」**

- jieba：`寫 / 不好` → 把「不好」留在一起，語意比較完整
- spaCy：`寫 / 不 / 好` → 拆成三個，失去了「寫不好」作為補語結構的語意

這個差異蠻有趣的，顯示不同工具對中文否定結構、補語的處理邏輯不太一樣。

---

**例子 4：「他知道這件事不要緊」**

- jieba 和 spaCy 都把「不要緊」切成「不要 / 緊」
- 但「不要緊」其實是固定用法（意思是「沒關係」），理想上應該保留為一個詞
- 這代表純詞典式或預設模型的斷詞，碰到慣用語時不一定切得對

---

另外值得注意的是，這批 sample data 裡有幾句本身就帶有歧義，像是「他知道這件事不要緊」「這份報告我寫不好」「我在樓上看到了他」「兒的生活好痛苦也沒有糧食多病少掙了很多錢」。這些句子因為斷詞位置不同，可能會產生不同的語意理解。例如最後一句完全沒有標點，讓兩套工具都很難處理，「病少」這種切法就明顯不對。

---

**整體來看：** 在這次測試中，jieba 的斷詞結果比 spaCy 更接近日常中文的使用習慣，spaCy 雖然用了深度學習模型，但 `zh_core_web_sm` 是比較小的模型，在繁體中文上的表現不太理想。不過兩套工具碰到專有名詞、口語固定搭配、歧義句或中英混合文字時，都還是需要人工確認結果。

## 7. Reflection

這次作業讓我比較深刻感受到中文 NLP 前處理跟英文真的差很多。英文有空格天然分隔，但中文的斷詞結果會直接影響後面的 POS tagging 和詞頻統計——前面切錯了，後面分析也會跟著歪掉。

像「台北市立美術館」這種專有名詞，或「不要緊」這種固定搭配，目前的工具都沒辦法完美處理。而且有些 sample data 裡的句子本身就有歧義（像「我在樓上看到了他」到底是「我在樓上」還是「他在樓上」），這種情況下不管用什麼工具斷詞，都很難保證結果是正確的，還是需要人自己判斷。

另外我原本以為 spaCy 一定比 jieba 強，但實際測試發現它的小模型（`zh_core_web_sm`）在繁體中文上表現並不好，反而 jieba 雖然只是基於詞典的統計方法，在多數情況下結果更合理。這讓我覺得選工具不能只看「技術比較新」，還是要看實際跑出來的結果。